### DISCUSSION TITLE
Is the Chamfer Distance (and knn) implementation of this Library really useful for my library ?

### DISCUSSION THREAD

**Discussion by scarpma** [facebookresearch/pytorch3d#1476] [Q&A]
Hello everybody and @nikhilaravi . First, congratulation for this very interesting and technically beautiful library. Great work !

I have developed a library for 3d surface mesh registration using pytorch and pytorch3d. I applied it to build a statistical shape model of the human aorta and we are publishing a paper about it (in which we cite your work, of course).

I started from [this](https://pytorch3d.org/tutorials/deform_source_mesh_to_target_mesh) tutorial to fit a mesh, but it didn't work as expected. I modified the gradient descent algorithm following this [work](https://rgl.epfl.ch/publications/Nicolet2021Large) and the results are really good. Basically, instead of adding a regularization term in the loss function, I add a preconditioning matrix (based on the laplacian of the source surface mesh) in front of the gradient of the loss function. This results in far better update steps that do not ruin the mesh during the optimization.

Since I would like to create a library with as few dependecies as possible, I would like to understand if the implementation of the Chamfer distance and knn of this library are really useful for my library. Since I don't use the batching mechanism, do you know other simpler implementations of the knn/chamfer distance in pytorch that would perform similarly for my case ?

Moreover, if you are interested, I could share parts of the core code (for now they are not allowing me to publish the full repository) to incorporate some examples in pytorch3d.
